src/io/trailer.cc

```diff
@@ -0,0 +1,156 @@
+    }
+    return checksum;
+}
+
+uint32_t
+GetFilesCheckSum(std::vector<std::string> files) {
+    uint32_t checksum = 0;
+    auto buffer = std::shared_ptr<uint8_t[]>(new uint8_t[kBlockSize]);
+    for (auto& file_name : files) {
+        std::ifstream reader(file_name.c_str(), std::ios::binary);
+        if (!reader) {
+            LOG_KNOWHERE_WARNING_ << file_name << "not exist, skip calculate check sum for this file.";
+            continue;
+        }
+        while (reader.read((char*)buffer.get(), kBlockSize)) {
+            std::streamsize read_size = reader.gcount();
+            checksum ^= CalculateCheckSum(buffer.get(), read_size);
+        }
+    }
+    return checksum;
+}
+}  // namespace
+
+namespace knowhere {
+void
+AddTrailerForMemoryIO(MemoryIOWriter& writer, const std::string& name, const Version& version) {
+    auto trailer_ptr = std::make_unique<Trailer>();
+    auto size = writer.tellg();
+    trailer_ptr->SetIndexBinarySize(size);
+    trailer_ptr->SetCheckSum(CalculateCheckSum(writer.data(), size));
+    trailer_ptr->SetVersion(version.VersionNumber());
+    trailer_ptr->SetIndexName(name);
+    writer.write(trailer_ptr->bytes, KNOWHERE_TRAILER_SIZE);
+}
+
+bool
+CheckTrailerForMemoryIO(MemoryIOReader& reader, const std::string& name) {
+    uint64_t bin_size = TRAILER_OFFSET(reader.size());
+    if (bin_size < 0) {
+        LOG_KNOWHERE_WARNING_ << "The binary is too small and assume no Trailer, pass Trailer check.";
+        return true;
+    }
+
+    auto trailer_ptr = std::make_unique<Trailer>();
+    auto pre_rp = reader.tellg();
+    reader.seekg(bin_size);
+    reader.read(trailer_ptr.get(), KNOWHERE_TRAILER_SIZE);
+    reader.seekg(pre_rp);
+
+    if (!trailer_ptr->TrailerValidCheck()) {
+        LOG_KNOWHERE_WARNING_ << "Trailer not exist in Binary.";
+        return true;
+    }
+
+    auto version = Version(trailer_ptr->GetVersion());
+    if (!Version::VersionSupport(version)) {
+        LOG_KNOWHERE_ERROR_ << "Index version(" << version.VersionNumber() << ") is not supported, Trailer check fail.";
+        return false;
+    }
+
+    if (trailer_ptr->GetIndexName() != name) {
+        LOG_KNOWHERE_ERROR_ << "Index type or data type is not correct(" << name << ").";
+        return false;
+    }
+
+    if (trailer_ptr->GetIndexBinarySize() != bin_size) {
+        LOG_KNOWHERE_ERROR_ << "The size of index binary is not correct.";
+        return false;
+    }
+
+    if (CalculateCheckSum(reader.data(), bin_size) != trailer_ptr->GetCheckSum()) {
+        LOG_KNOWHERE_ERROR_ << "Binary checksum check fail.";
+        return false;
+    }
+    LOG_KNOWHERE_INFO_ << "Index Trailer check succeed.";
+    return true;
+}
+
+void
+AddTrailerForFiles(const std::vector<std::string>& files, const std::string& trailer_file, const std::string& name,
+                   const Version& version) {
+    auto trailer_ptr = std::make_unique<Trailer>();
+    trailer_ptr->SetCheckSum(GetFilesCheckSum(files));
```

Comment:
trailer for files is missing binary size. I prefer to make them consistent:

* either remove binary size from single binary index and rely solely on checksum for correctness, which should be sufficient?
* or add binary size sum of all files for multiple files index

Reply:
binary size removed.